// @sap/generator-cds/lib/helper/term.js
const debug = process.env.DEBUG

// https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences
const term = {
  reset: '\x1b[0m', // Default
  bold: '\x1b[1m', // Bold/Bright
  link: '\x1b[4m', // underline
  red: '\x1b[91m', // Bright Foreground Red
  green: '\x1b[32m', // Foreground Green
  orange: '\x1b[38;2;255;140;0m' // darker orange, works with bright and dark background
}

const asErr = module.exports.error = o => debug ? o : as(term.red + term.bold, o)
const asWarn = module.exports.warn = o => debug ? o : as(term.orange + term.bold, o)
const asInfo = module.exports.info = o => debug ? o : as(term.green + term.bold, o)
module.exports.warn  = o => as(term.orange, o)
module.exports.info  = o => as(term.green, o)
module.exports.link  = o => as(term.link, o)
Object.defineProperty(module.exports, 'colors', { get: () => process.stdin.isTTY && process.stdout.isTTY })

const format = module.exports.format = (o, severity='Error', asInternalError=false) => {
  switch (severity) {
    case 'Error'  : return format.error (o, asInternalError)
    case 'Warning': return format.warn (o)
    default       : return format.info (o)
  }
}

// decorate.error, .warning, .info
// 'Error: foo'  ->  '[ERROR] foo'  (Maven-like, allows for better grepping in logs)
Object.assign (format, {
  error: (o, asInternalError) => {
    if (debug)  return o
    if (asInternalError) {
      return `[${asErr('INTERNAL ERROR')}] ${o.stack || o.toString()}\nPlease report this error.\n`
    }
    return `[${asErr('ERROR')}] ${toString(o, 'Error')}`
  },
  warn: o => debug ? o : `[${asWarn('WARNING')}] ${toString(o, 'Warning')}`,
  info: o => debug ? o : `[${asInfo('INFO')}] ${toString(o, 'Info')}`
})

function as(codes, o) {
  return module.exports.colors ? (codes + o + term.reset) : ('' + o)
}

function toString(o, severity) { //NOSONAR
  if (!o || !o.toString)  return o
  // strips the 'Error: ' prefix in the message, so that we can add our own prefix
  return o.toString()
      .replace(new RegExp('^' + severity + ': ', 'i'), '')  // beginning
      .replace(new RegExp(' ' + severity + ':' , 'i'), '')  // middle
}
