// @sap/generator-cds/lib/constants.js
const constants = {

    FILE_NAME_CDSRC_JSON: '.cdsrc.json',
    FILE_NAME_PACKAGE_JSON: 'package.json',

    ENV_KEY_CDS_IDE: '_CDS_IDE',
    ENV_KEY_CDS_IDE_WEBIDE: 'webide',
    ENV_KEY_CDS_IDE_ECLIPSE: 'eclipse',

    ENV_VAR_GENERATION_LOG: 'CDS_BUILD_OUTPUTFILE',

    // add new module type here
    MODULE_TYPE_DB: 'db',
    MODULE_TYPE_SRV: 'srv',

    DB_TECHNOLOGY_HANA: 'hana',
    DB_TECHNOLOGY_SQLITE: 'sqlite',

    PRIMARY_DB: 'db',
    HDI_CONTAINER_TYPE: 'com.sap.xs.hdi-container',

    SRV_TECHNOLOGY_JAVA: 'java',
    SRV_TECHNOLOGY_NODEJS: 'nodejs',

    DEFAULT_GEN_FOLDER: 'gen',

    DEFAULT_MODULE_FOLDER: {},

    TASKS: {},

    RESOURCE_TYPE_UAA: 'com.sap.xs.uaa',
    RESOURCE_TYPE_MANAGED_SERVICE: 'org.cloudfoundry.managed-service',
    SERVICE_TYPE_XSUAA: 'xsuaa',

    UAA_SERVICE_PLAN_APPLICATION: 'application',
    UAA_SERVICE_PLAN_DEFAULT: 'default',

    WINDOWS_FORBIDDEN_FOLDER_NAMES: ['CON', 'PRN', 'AUX', 'NUL',
        'COM1', 'COM2', 'COM3', 'COM4', 'COM5', 'COM6', 'COM7', 'COM8', 'COM9',
        'LPT1', 'LPT2', 'LPT3', 'LPT4', 'LPT5', 'LPT6', 'LPT7', 'LPT8', 'LPT9'],

    ODATA_VERSION_V2: 'v2',
    ODATA_VERSION_V4: 'v4',
    DEFAULT_ODATA_VERSION: 'v2',

    CLOUD_SDK_GROUP_ID: 'com.sap.cloud.servicesdk.prov',

    REGEX_JAVA_PACKAGE: /^[a-zA-Z]\w*(\.[a-zA-Z]\w*)*$/g,
    REGEX_SRV_MEMORY: /^[1-9]\d*\w?$/,
    REGEX_CDS_VERSION: /^([~^]?(\d+|\*)(\.(\d+|\*)){0,2}|latest)$/,

    SAP_HANA_CLIENT: '@sap/hana-client'
};

// add new module type here
constants.TASKS[constants.MODULE_TYPE_DB] = {};
constants.TASKS[constants.MODULE_TYPE_DB][constants.DB_TECHNOLOGY_HANA] = 'hana';

constants.TASKS[constants.MODULE_TYPE_SRV] = {};
constants.TASKS[constants.MODULE_TYPE_SRV][constants.SRV_TECHNOLOGY_JAVA] = 'java-cf';
constants.TASKS[constants.MODULE_TYPE_SRV][constants.SRV_TECHNOLOGY_NODEJS] = 'node-cf';

// add new module type here
constants.DEFAULT_MODULE_FOLDER[constants.MODULE_TYPE_DB] = 'db';
constants.DEFAULT_MODULE_FOLDER[constants.MODULE_TYPE_SRV] = 'srv';

module.exports = constants;
